File: logger.py
from enum import IntEnum
import json
import threading
import typing

import time
import zmq

ZMQ_PORT = 5556
ZMQ_TIMEOUT_MS = 100

log_manager = None

start_time = time.time()

# Would be replaced by PLib_GetTime call
def time_now():
    return (time.time() - start_time)

# LogLevel has two (related, but different) applications:
#  1. Allow loggers to specify the severity of messages
#  2. Sinks use them to filter messages, based on how the subscription was configured
# Perhaps the latter use case could be extracted into a different "Filter" class, allowing
# sinks to more accurately filter what logs they are interested in and removing the need for "ANY"
class LogLevel(IntEnum):
    ANY      = 0
    DEBUG    = 1
    INFO     = 2
    WARNING  = 3
    ERROR    = 4
    CRITICAL = 5

    def pretty_name(self, with_color = False):
        name = {
            LogLevel.ANY     : "Not Set",
            LogLevel.DEBUG   : "Debug",
            LogLevel.INFO    : "Info",
            LogLevel.WARNING : "Warning",
            LogLevel.ERROR   : "Error",
            LogLevel.CRITICAL: "CRITICAL",
        }[self]

        if with_color:
            color = {
                LogLevel.ANY     : "\033[97m",
                LogLevel.DEBUG   : "\033[96m",
                LogLevel.INFO    : "\033[92m",
                LogLevel.WARNING : "\033[93m",
                LogLevel.ERROR   : "\033[91m",
                LogLevel.CRITICAL: "\033[31m",
            }[self]
            name = color + name + "\033[0m"

        return ('[' + name + ']')

class Log():
    def __init__(self, source: [str], time: float, level: LogLevel, message: str):
        self.source = source
        self.time = time
        self.level = level
        self.message = message

    def from_json(json_str: str):
        d = json.loads(json_str)
        return Log(d["source"], d["time"], LogLevel(d["level"]), d["message"])

    def to_json(self):
        return json.dumps(self.__dict__)

class LogSink():
    def __init__(self):
        pass
        
    # def subscribe(self, logger_name: str, level: LogLevel, medium = LogMedium):
    #     # source.add_subscriber(level, self.receive_log)
    #     medium.subscribe(self, logger, level)

    def receive_log(self, log):
        pass

    def subscribe(self, logger_name: str, level: LogLevel, medium = "default"):
        log_manager.mediums[medium].subscribe(self, logger_name, level)

    def unsubscribe(self, logger_name: str, medium = "default"):
        log_manager.mediums[medium].unsubscribe(self, logger_name)

class Logger(LogSink):
    def __init__(self, manager, name: str):
        LogSink.__init__(self)
        self.manager = manager
        self.name = name
        # self.callbacks = [[] for _ in LogLevel]
        self.callbacks = []

    def add_subscriber(self, level: LogLevel, callback):
        self.callbacks.append(callback)

    # Pass on the logs from below, adding own name to the source
    def receive_log(self, log):
        log.source.insert(0, self.name)
        self.publish_log(log)

    def publish_log(self, log):
        for callback in self.callbacks:
            callback(log)

    def equip(self, medium: str):
        log_manager.mediums[medium].equip(self)

    # User commands for creating logs
    def debug(self, message: str):
        self.publish_log(Log([self.name], time_now(), LogLevel.DEBUG, message))
    def info(self, message: str):
        self.publish_log(Log([self.name], time_now(), LogLevel.INFO, message))
    def warning(self, message: str):
        self.publish_log(Log([self.name], time_now(), LogLevel.WARNING, message))
    def error(self, message: str):
        self.publish_log(Log([self.name], time_now(), LogLevel.ERROR, message))
    def critical(self, message: str):
        self.publish_log(Log([self.name], time_now(), LogLevel.CRITICAL, message))

class LogMedium():
    def __init__(self):
        self.loggers = []

    # Register a new logger in the medium
    def add_logger(self, logger: Logger):
        self.loggers += logger
        pass

    # Publish a log from a given logger 
    def publish_log(self, logger, log):
        pass

    # Connect a logger to a sink
    def connect(self, source: Logger, sink: LogSink):
        pass

class LogMediumZMQ(LogMedium):
    def __init__(self, port: int):
        super().__init__()
        self.ctx = zmq.Context()

        self.pub_socket = self.ctx.socket(zmq.PUB)
        self.pub_socket.bind("tcp://*:{}".format(port))

        self.sub_socket = self.ctx.socket(zmq.SUB)
        self.sub_socket.connect("tcp://localhost:{}".format(port))

        # Create a poller to poll messages
        self.poller = zmq.Poller()
        self.poller.register(self.sub_socket, zmq.POLLIN)

        self.zmq_lock = threading.Lock()

        self.subscription_thread = None
        self.exist_subscriptions = False # Do there exist any subscriptions?

        # Dictionary of logger name to dict sink to level
        self.subscriptions = {}

    def stop(self):
        self.exist_subscriptions = False
        self.subscription_thread.join()

    # Function that updates all sinks once new 
    def subscription_handler(self):
        while self.exist_subscriptions:
            self.propagate(timeout = ZMQ_TIMEOUT_MS)

    def propagate(self, timeout = ZMQ_TIMEOUT_MS):
        while self.sub_socket in dict(self.poller.poll(timeout)):
            log_topic = None
            log_json = None

            # We need to check that there is still stuff to read after acquiring the lock
            # since someone could have stolen our stuff in between! (It happens a lot otherwise)
            with self.zmq_lock:
                if self.sub_socket in dict(self.poller.poll(timeout)):
                    log_topic = bytes.decode(self.sub_socket.recv())
                    log_json = bytes.decode(self.sub_socket.recv())

            if not (log_topic and log_json):
                continue

            # [log_topic, log_json] = list(map(bytes.decode, multipart_resp))
            log = Log.from_json(log_json)

            for (sink, level) in self.subscriptions[log_topic].items():
                # Give log to sink if it is of a high enough level
                if log.level >= level:
                    sink.receive_log(log)

    def publish_log(self, log):
        # Propagate messages if any exist, but do not wait if there are none
        self.propagate(timeout = 0)

        with self.zmq_lock:
            self.pub_socket.send_multipart([
                log.source[0].encode('ascii'),
                log.to_json().encode('ascii')
            ])

    def equip(self, logger: Logger):
        logger.callbacks.append(self.publish_log)

    def subscribe(self, sink: LogSink, logger_name: str, level: LogLevel):
        if logger_name not in self.subscriptions:
            self.subscriptions[logger_name] = {}
            self.sub_socket.setsockopt_string(zmq.SUBSCRIBE, logger_name)
        self.subscriptions[logger_name][sink] = level

        if not self.subscription_thread:
            self.exist_subscriptions = True
            self.subscription_thread = threading.Thread(target = self.subscription_handler)
            self.subscription_thread.start()

    def unsubscribe(self, sink: LogSink, logger_name: str):
        if logger_name not in self.subscriptions:
            return

        self.subscriptions[logger_name].pop(sink, None)

        # If the topic has no more subscribers, stop listening for it on the socket
        if not self.subscriptions[logger_name]:
            self.sub_socket.setsockopt_string(zmq.UNSUBSCRIBE, logger_name)
            self.subscriptions.pop(logger_name)

        # Check if there are any subscriptions whatsoever
        if not self.subscriptions:
            self.exist_subscriptions = False

class LogManager():
    def __init__(self, config = None):
        self.loggers = {}
        if config:
            with open(config) as json_conf:
                self.conf = json.load(json_conf)
        else:
            self.conf = {}

        self.mediums = {
            "default": LogMedium(),
            "zmq": LogMediumZMQ(ZMQ_PORT),
        }

    def getLogger(self, name: str):
        # Return if it has been initialised, otherwise initialise it first
        if name in self.loggers:
            return self.loggers[name]
        else:
            logger = Logger(self, name)
            self.loggers[name] = logger
            
            # if name in self.conf:
            #     for medium_name, medium_conf in self.conf[name]:
            #         pass

            return logger

# Different types of sinks/loggers 
class LogSinkPrinter(LogSink):
    def __init__(self, with_color = True):
        self.with_color = with_color
        self.max_tag_length = max([
            len(level.pretty_name(with_color = with_color)) for level in LogLevel
        ])

    def receive_log(self, log):
        print("{time:9.5f} {tag:<{tag_len}} {message} ({source})".format(
            time = log.time,
            tag = log.level.pretty_name(with_color = self.with_color),
            tag_len = self.max_tag_length,
            message = log.message,
            source = '.'.join(log.source)
        ))

log_manager = LogManager()

def getLogger(name: str):
    return log_manager.getLogger(name)

if __name__ == "__main__":
    child_logger = getLogger("child")
    child_logger.equip("zmq")
    
    logger = getLogger("main")
    logger.equip("zmq")
    logger.subscribe("child", LogLevel.WARNING, medium = "zmq")

    printer = LogSinkPrinter(with_color = True)
    printer.subscribe("main", LogLevel.ANY, medium = "zmq")

    # child_printer = LogSinkPrinter()
    # child_printer.subscribe("child", LogLevel.ANY, medium = "zmq")

    logger.info("This is just some info! 1")
    logger.info("This is just some info! 2")
    logger.info("This is just some info! 3")
    logger.debug("This is some good debug info 1")
    logger.info("This is just some info! 4")
    logger.info("This is just some info! 5")
    logger.debug("This is some good debug info 2")

    child_logger.info("Everything is good for now! :)")
    child_logger.warning("Oopsies!")
    # time.sleep(0.001)
    child_logger.error("It got worse!")

    # time.sleep(0.001)

    logger.critical("Shutting down!")

    time.sleep(1)

    # This causes issues with the logger currently if unsubscription happens
    # while logs are being sent out
    logger.unsubscribe("child", medium = "zmq")
    printer.unsubscribe("main", medium = "zmq")
    # child_printer.unsubscribe("child", medium = "zmq")

    # !!! There is also the issue that sometimes logs coming from children are published after logs coming from parent, even if the child log came in before

    # childlogger = Logger("child")

    # logger = Logger("main")
    # logger.subscribe(childlogger, LogLevel.ANY)

    # printer = LogSinkPrinter(with_color = True)
    # printer.subscribe(childlogger, LogLevel.ANY)

    # logger.info("This is just some info!")
    # logger.info("This is just some info!")
    # logger.info("This is just some info!")
    # logger.debug("This is some good debug info")
    # logger.info("This is just some info!")
    # logger.info("This is just some info!")
    # logger.debug("This is some good debug info")

    # childlogger.warning("Oopsies!")
    # childlogger.error("It got worse!")

    # logger.critical("Shutting down!")

File: logger.cpp
#include <chrono>
#include <iostream>

#include "logger.hpp"

using namespace std::chrono;

auto time_start = high_resolution_clock::now();

double time_now() {
    return duration<double>(high_resolution_clock::now() - time_start).count();
}

namespace logging {

string LogLevel_toStr(LogLevel level, bool with_color) {
    string name;
    switch(level) {
        case ANY:      {name = "ANY"; break;}
        case DEBUG:    {name = "DEBUG"; break;}
        case INFO:     {name = "INFO"; break;}
        case WARNING:  {name = "WARNING"; break;}
        case ERROR:    {name = "ERROR"; break;}
        case CRITICAL: {name = "CRITICAL"; break;}
    }

    if(with_color) {
        string color_code;
        switch(level) {
            case ANY:      {color_code = "\033[97m"; break;}
            case DEBUG:    {color_code = "\033[96m"; break;}
            case INFO:     {color_code = "\033[92m"; break;}
            case WARNING:  {color_code = "\033[93m"; break;}
            case ERROR:    {color_code = "\033[91m"; break;}
            case CRITICAL: {color_code = "\033[31m"; break;}
        }

        return color_code + name + "\033[0m";
    }

    return name;
}

Log::Log(string source, double time, LogLevel level, string message) {
    this->source = source;
    this->time = time;
    this->level = level;
    this->message = message;
}

void Sink::subscribe(Logger *logger) {
    logger->attach_sink(weak_from_this());
}

void Sink::subscribe(string logger) {
    subscribe(get(logger));
}

void Sink::unsubscribe(Logger *logger) {
    logger->deattach_sink(weak_from_this());
}

void Sink::unsubscribe(string logger) {
    unsubscribe(get(logger));
}

Logger::Logger(Logger *parent, string name, bool propagate) {
    this->parent = parent; // If parent is a null pointer, it is the root logger
    this->name = name;
    this->propagate = propagate;

    // self.children = {}
    // self.sinks = WeakSet()
}

void Logger::publish_log(Log log) {
    // std::vector<std::weak_ptr<Sink>>::iterator iter;
    // for(iter = sinks.begin(); iter != sinks.end(); ) {
    //     if(auto sink_ptr = (*iter).lock()) {
    //         sink_ptr->handle(log);
    //         ++iter;
    //     } else {
    //         iter = sinks.erase(iter);
    //     }
    // }

    for(std::weak_ptr<Sink> weak_ptr : sinks) {
        if(auto sink_ptr = weak_ptr.lock()) {
            sink_ptr->handle(log);
        } else {
            sinks.erase(weak_ptr);
        }
    }

    if(parent and propagate) {
        parent->publish_log(log);
    }
}

void Logger::attach_sink(std::weak_ptr<Sink> sink) {
    sinks.insert(sink);
}

void Logger::deattach_sink(std::weak_ptr<Sink> sink) {
    sinks.erase(sink);
}

void Logger::ensure_child(string logger_name) {
    if(children.find(logger_name) == children.end()) {
        children.try_emplace(logger_name, this, name + "/" + logger_name);
    }
}

Logger *Logger::get(string logger_name) {
    int delim_index = logger_name.find("/");

    string head = logger_name.substr(0, delim_index);
    ensure_child(head);

    if(delim_index == string::npos) {
        auto iter = children.find(head);
        return iter != children.end() ? &iter->second : nullptr;
    } else {
        string tail = logger_name.erase(0, delim_index + 1);

        auto iter = children.find(head);
        return iter != children.end() ? (&iter->second)->get(tail) : nullptr;
    }
}

void Logger::debug(string message) { publish_log(Log(name, time_now(), LogLevel::DEBUG, message)); }
void Logger::info(string message) { publish_log(Log(name, time_now(), LogLevel::INFO, message)); }
void Logger::warning(string message) { publish_log(Log(name, time_now(), LogLevel::WARNING, message)); }
void Logger::error(string message) { publish_log(Log(name, time_now(), LogLevel::ERROR, message)); }
void Logger::critical(string message) { publish_log(Log(name, time_now(), LogLevel::CRITICAL, message)); }

Logger *get(string logger_name) {return root_logger.get(logger_name);}

namespace sinks {

using std::string;
PrintSink::PrintSink(bool with_color) {
    this->with_color = with_color;
}

void PrintSink::handle(logging::Log log) {
    std::cout << std::setw(9) << std::setprecision(5) << std::fixed << log.time << " ";
    std::cout << "[" << LogLevel_toStr(log.level, with_color = this->with_color) << "] ";
    std::cout << log.message << " ";
    std::cout << "(" << log.source.substr(1) << ")" << std::endl;
}

FileSink::FileSink(string file_path) {
    using ios = std::ios;
    file_stream.open(file_path, ios::out | ios::app);
}

FileSink::~FileSink() {
    file_stream.close();
}

void FileSink::handle(logging::Log log) {
    file_stream << std::setw(9) << std::setprecision(5) << std::fixed << log.time << " ";
    file_stream << "[" << LogLevel_toStr(log.level) << "] ";
    file_stream << log.message << " ";
    file_stream << "(" << log.source.substr(1) << ")" << std::endl;
}

using nlohmann::json;

ZMQPubSink::ZMQPubSink(string host, unsigned int port, string topic): socket{ctx, ZMQ_PUB} {
    std::ostringstream ss;
    ss << "tcp://" << host << ":" << port;
    this->address = ss.str();
    this->topic = topic;
    socket.bind(address);
}

void ZMQPubSink::handle(logging::Log log) {
    json log_json;
    to_json(log_json, log);
    string log_str = log_json.dump();

    std::vector<zmq::message_t> msgs;

    msgs.emplace_back(topic.c_str(),
                      topic.size());

    msgs.emplace_back(log_str.c_str(),
                      log_str.size());

    zmq::send_multipart(socket, msgs);
}
}
}